rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,977 @@
+      this.pendingCacheEntry = pendingCacheEntry;
+      this.backoffCacheEntry = backoffCacheEntry;
+      checkState((dataCacheEntry != null ^ pendingCacheEntry != null ^ backoffCacheEntry != null)
+          && !(dataCacheEntry != null && pendingCacheEntry != null && backoffCacheEntry != null),
+          "Expected only 1 cache entry value provided");
+    }
+
+    static CachedRouteLookupResponse pendingResponse(PendingCacheEntry pendingEntry) {
+      return new CachedRouteLookupResponse(pendingEntry.request, null, pendingEntry, null);
+    }
+
+    static CachedRouteLookupResponse backoffEntry(BackoffCacheEntry backoffEntry) {
+      return new CachedRouteLookupResponse(backoffEntry.request, null, null, backoffEntry);
+    }
+
+    static CachedRouteLookupResponse dataEntry(DataCacheEntry dataEntry) {
+      return new CachedRouteLookupResponse(dataEntry.request, dataEntry, null, null);
+    }
+
+    boolean hasData() {
+      return dataCacheEntry != null;
+    }
+
+    @Nullable
+    ChildPolicyWrapper getChildPolicyWrapper() {
+      if (!hasData()) {
+        return null;
+      }
+      return dataCacheEntry.getChildPolicyWrapper();
+    }
+
+    @Nullable
+    public String getHeaderData() {
+      if (!hasData()) {
+        return null;
+      }
+      return dataCacheEntry.getHeaderData();
+    }
+
+    boolean hasError() {
+      return backoffCacheEntry != null;
+    }
+
+    boolean isPending() {
+      return pendingCacheEntry != null;
+    }
+
+    @Nullable
+    Status getStatus() {
+      if (!hasError()) {
+        return null;
+      }
+      return backoffCacheEntry.getStatus();
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("request", request)
+          .add("dataCacheEntry", dataCacheEntry)
+          .add("pendingCacheEntry", pendingCacheEntry)
+          .add("backoffCacheEntry", backoffCacheEntry)
+          .toString();
+    }
+  }
+
+  /** A pending cache entry when the async RouteLookup RPC is still on the fly. */
+  final class PendingCacheEntry {
+    private final ListenableFuture<RouteLookupResponse> pendingCall;
+    private final RouteLookupRequest request;
+    private final BackoffPolicy backoffPolicy;
+
+    PendingCacheEntry(
+        RouteLookupRequest request, ListenableFuture<RouteLookupResponse> pendingCall) {
+      this(request, pendingCall, null);
+    }
+
+    PendingCacheEntry(
+        RouteLookupRequest request,
+        ListenableFuture<RouteLookupResponse> pendingCall,
+        @Nullable BackoffPolicy backoffPolicy) {
+      this.request = checkNotNull(request, "request");
+      this.pendingCall = pendingCall;
+      this.backoffPolicy = backoffPolicy == null ? backoffProvider.get() : backoffPolicy;
+      pendingCall.addListener(
+          new Runnable() {
+            @Override
+            public void run() {
+              handleDoneFuture();
+            }
+          },
+          synchronizationContext);
```

Comment:
Why only this place is using `synchronizationContext`, whereas all other places are using `scheduledExecutorService`? Is there any guidance when to use which?

Reply:
good catch! both places need to use synchronizationContext especially if it can transition to DataCacheEntry.